File: crates/controller/src/interface/http/storage.rs
use axum::{
    Json,
    extract::{Query, State},
    response::Response,
};
use chrono::{DateTime, Utc};
use switchboard_custom_config::SerdeValue;
use switchboard_model::{
    FlattenPageQueryWithFilter, PagedList, ServiceConfig, error::ResultObject,
};

use crate::{
    interface::http::HttpState,
    storage::{ListObjectQuery, ObjectFilter, StorageObjectDescriptor, StorageObjectWithoutData},
};

#[derive(Debug, serde::Deserialize)]
pub struct SaveStorageObjectRequest {
    pub id: String,
    pub date_type: String,
    pub data: SerdeValue,
}

#[derive(Debug, serde::Deserialize)]
#[serde(tag = "mode", rename_all = "snake_case")]
pub enum BatchDeleteStorageObjectRequest {
    ByDescriptors {
        descriptors: Vec<StorageObjectDescriptor>,
    },
    ById {
        id: String,
    },
}

pub async fn get(
    State(state): State<HttpState>,
    Query(params): Query<StorageObjectDescriptor>,
) -> Response {
    let process = async {
        let obj = state
            .controller_context
            .storage()
            .get_object(&params)
            .await?;
        crate::Result::Ok(obj)
    };
    super::result_to_json_response(process.await)
}

pub async fn delete(
    State(state): State<HttpState>,
    Query(params): Query<StorageObjectDescriptor>,
) -> Response {
    let process = async {
        state
            .controller_context
            .storage()
            .delete_object(&params)
            .await?;
        crate::Result::Ok(())
    };
    super::result_to_json_response(process.await)
}

pub async fn batch_delete(
    State(state): State<HttpState>,
    Json(request): Json<BatchDeleteStorageObjectRequest>,
) -> Response {
    let process = async {
        match request {
            BatchDeleteStorageObjectRequest::ByDescriptors { descriptors } => {
                state
                    .controller_context
                    .storage()
                    .batch_delete_objects(descriptors)
                    .await?;
            }
            BatchDeleteStorageObjectRequest::ById { id } => {
                state
                    .controller_context
                    .storage()
                    .delete_all_objects_by_id(&id)
                    .await?;
            }
        }
        crate::Result::Ok(())
    };
    super::result_to_json_response(process.await)
}

pub async fn save(
    State(state): State<HttpState>,
    Json(request): Json<SaveStorageObjectRequest>,
) -> Response {
    let process = async {
        let descriptor = state
            .controller_context
            .storage()
            .save_object(&request.id, &request.date_type, request.data)
            .await?;
        crate::Result::Ok(descriptor)
    };
    super::result_to_json_response(process.await)
}

#[derive(Debug, serde::Serialize, serde::Deserialize, Clone)]
pub struct ObjectFilterQuery {
    pub data_type: Option<String>,
    pub id: Option<String>,
    pub revision: Option<String>,
    pub latest_only: Option<BooleanLit>,
    pub created_before: Option<DateTime<Utc>>,
    pub created_after: Option<DateTime<Utc>>,
}

#[derive(Debug, serde::Serialize, serde::Deserialize, Clone)]
#[serde(rename_all = "snake_case")]
pub enum BooleanLit {
    True,
    False,
}

impl From<BooleanLit> for bool {
    fn from(value: BooleanLit) -> Self {
        match value {
            BooleanLit::True => true,
            BooleanLit::False => false,
        }
    }
}

pub async fn list(
    State(state): State<HttpState>,
    Query(params): Query<FlattenPageQueryWithFilter<ObjectFilterQuery>>,
) -> Response {
    let (page, filter) = params.into_parts();
    let filter: ObjectFilter = ObjectFilter {
        data_type: filter.data_type,
        id: filter.id,
        revision: filter.revision,
        latest_only: filter.latest_only.map(|v| v.into()),
        created_before: filter.created_before,
        created_after: filter.created_after,
    };
    let process = async {
        let list = state
            .controller_context
            .storage()
            .list_objects(ListObjectQuery { page, filter })
            .await?;
        crate::Result::Ok(list)
    };
    super::result_to_json_response(process.await)
}

pub fn router() -> axum::Router<HttpState> {
    axum::Router::new()
        .route("/object", axum::routing::get(get).delete(delete).post(save))
        .route("/objects", axum::routing::get(list).delete(batch_delete))
}

File: ui/controller-web/src/lib/api/routes/storage.ts
import { fetchJson, fetchQuery } from '.';
import type { StorageObjectDescriptor, StorageObjectWithoutData } from '../types';
import type { PagedResult, PageQuery } from '../types/paged';

export type SaveStorageObjectRequest = {
	resolver: string;
	save_as?: string;
	config: unknown;
};

export type BatchDeleteStorageObjectRequest =
	| {
			mode: 'by_descriptor';
			objects: StorageObjectDescriptor[];
	  }
	| {
			mode: 'by_id';
			id: string;
	  };

export type ObjectFilter = {
	data_type?: string;
	id?: string;
	revision?: string;
	latest_only?: boolean;
	created_before?: Date;
	created_after?: Date;
};

export const storageApi = {
	get: (descriptor: StorageObjectDescriptor) =>
		fetchQuery<unknown>('/api/storage/object', new URLSearchParams(descriptor)),
	save: (request: SaveStorageObjectRequest) =>
		fetchJson<StorageObjectDescriptor>('/api/storage/object', {
			method: 'POST',
			body: JSON.stringify(request)
		}),
	delete: (descriptor: StorageObjectDescriptor) =>
		fetchQuery<void>('/api/storage/object', new URLSearchParams(descriptor), {
			method: 'DELETE'
		}),
	list(page: PageQuery, filter: ObjectFilter) {
		const params = new URLSearchParams({
			limit: page.limit.toString(),
			...(page.cursor.next ? { next: page.cursor.next } : {}),
			...(filter.data_type ? { data_type: filter.data_type } : {}),
			...(filter.id ? { id: filter.id } : {}),
			...(filter.revision ? { revision: filter.revision } : {}),
			...(filter.latest_only !== undefined ? { latest_only: filter.latest_only.toString() } : {}),
			...(filter.created_before ? { created_before: filter.created_before.toISOString() } : {}),
			...(filter.created_after ? { created_after: filter.created_after.toISOString() } : {})
		});
		return fetchQuery<PagedResult<StorageObjectWithoutData>>('/api/storage/objects', params);
	},
	batchDelete: (request: BatchDeleteStorageObjectRequest) =>
		fetchJson<void>('/api/storage/objects', {
			method: 'POST',
			body: JSON.stringify(request)
		})
};
